sdks/python/apache_beam/transforms/external.py

```diff
@@ -144,6 +147,101 @@ def _get_named_tuple_instance(self):
+  """
+  def __init__(self, class_name):
+    """
+    :param class_name: fully qualified name of the transform class.
+    """
+    if not class_name:
+      raise ValueError('Class name must not be empty')
+
+    self._class_name = class_name
+    self._constructor_method = None
+    self._constructor_params = None
+    self._builder_methods_and_params = OrderedDict()
+
+  def _get_schema_proto_and_payload(self, **kwargs):
+    named_fields = []
+    for key, value in kwargs.items():
+      if not key:
+        raise ValueError('Parameter name cannot be empty')
+      if value is None:
+        raise ValueError(
+            'Received value None for key %s. None values are currently not '
+            'supported' % key)
+      named_fields.append(
+          (key, convert_to_typing_type(instance_to_type(value))))
+
+    schema_proto = named_fields_to_schema(named_fields)
+    row = named_tuple_from_schema(schema_proto)(**kwargs)
+    schema = named_tuple_to_schema(type(row))
+
+    payload = RowCoder(schema).encode(row)
+    return (schema_proto, payload)
+
+  def build(self):
+    constructor_params = self._constructor_params or {}
+    constructor_schema, constructor_payload = (
+        self._get_schema_proto_and_payload(**constructor_params))
+    payload = JavaClassLookupPayload(
+        class_name=self._class_name,
+        constructor_schema=constructor_schema,
+        constructor_payload=constructor_payload)
+    if self._constructor_method:
+      payload.constructor_method = self._constructor_method
+
+    for builder_method_name, params in self._builder_methods_and_params.items():
+      builder_method_schema, builder_method_payload = (
+          self._get_schema_proto_and_payload(**params))
+      builder_method = BuilderMethod(
+          name=builder_method_name,
+          schema=builder_method_schema,
+          payload=builder_method_payload)
+      builder_method.name = builder_method_name
+      payload.builder_methods.append(builder_method)
+    return payload
+
+  def add_constructor(self, **kwargs):
+    """
+    Specifies the Java constructor to use.
+
+    :param kwargs: parameter names and values of the constructor.
+    """
+    if self._constructor_method or self._constructor_params:
+      raise ValueError(
+          'Constructor or constructor method can only be specified once')
+
+    self._constructor_params = kwargs
+
+  def add_constructor_method(self, method_name, **kwargs):
+    """
+    Specifies the Java constructor method to use.
+
+    :param method_name: name of the constructor method.
+    :param kwargs: parameter names and values of the constructor method.
+    """
+    if self._constructor_method or self._constructor_params:
+      raise ValueError(
+          'Constructor or constructor method can only be specified once')
+
+    self._constructor_method = method_name
+    self._constructor_params = kwargs
+
+  def add_builder_method(self, method_name, **kwargs):
```

Comment:
Here (and elsewhere) should we accept args as well as kwargs, as java arguments are unnamed?

Reply:
I think there are multiple issues with supporting that.

(1) Java args have to be ordered but Python requires args to be specified before kwargs. So we'll somehow have to preserve the order between args and kwargs or accept only one. That will probably end up being confusing to users.
(2) Schema requires field names to be specified. So we'll have to define a special set of field names that get ignored by Java side.

Also, Java parameter names should be already known (for example, available in Java docs) and it should not be too difficult for Python users to specify these.

WDYT ?

Reply:
Kwargs always come after args, so the ordering between them is well-defined. 

One issue is that the java argument names are not part of the signature's spec, so may not be stable, or may not even be in the .class files at all. As for defining names, a simple scheme that uses "illegal" parameter names (e.g. `#N`) should be sufficient. 

Do we correctly handle `builderMethod(a=1, b=2)` vs. `builderMethod(b=2, a=1)` or otherwise enforce the names today? If not, perhaps we should only allow positional arguments, as one would expect those two to be exactly the same.

Reply:
> Ordering

Right. The issue goes away if we always use *args before *kwargs (in the order they are specified).

> One issue is that the java argument names are not part of the signature's spec

This should be fine. We ignore argument names at Java side when the are not available through reflection. 

> As for defining names, a simple scheme that uses "illegal" parameter names (e.g. #N) should be sufficient.

Sounds good I can add that and update the spec. I assume you meant that certain parameter names that will be ignored by the expansion service and will not be used for validation.

> Do we correctly handle builderMethod(a=1, b=2) vs. builderMethod(b=2, a=1) or otherwise enforce the names today?

We cannot reorder args since that could lead to ambiguities in Java side. Hence we send the args exactly in the order they are specified. May be we can clarify this in the documentation ?

Reply:
Note that Python 3.6 and above preserve the order of **kwargs: https://www.python.org/dev/peps/pep-0468/

Reply:
Added support for sending arbitrary field names through a well defined pattern that will be ignored.

Added support for positional args in Python side using above.

Updated Python documentation to clarify that args and kwargs for constructor/builder methods will be applied in the specified order.

Reply:
Thanks. 

Just to confirm, if the Java constructor/method does not have an argument named `foo` then `builderMethod(foo=x)` will fail, right? And if it does have an argument named `foo` and `bar` then `builderMethod(foo=x, bar=y)` will produce the same result as `builderMethod(bar=y, foo=x)`?

Reply:
> if the Java constructor/method does not have an argument named foo then builderMethod(foo=x) will fail, right?

Yes (unless arg name starts with the predefined 'ignore' prefix). Also we can only perform validation if Java reflection preserved arg names otherwise we just validate using types.

> And if it does have an argument named foo and bar then builderMethod(foo=x, bar=y) will produce the same result as builderMethod(bar=y, foo=x) (or one of them rejected as the name order doesn't match)?

If Java builder method is 'myMethod(int foo, int bar)'

Then 'builderMethod(foo=x, bar=y)' will pass while 'builderMethod(bar=x, foo=y)' will fail. In other words, Python arguments have to be specified in exactly the same order as Java. I documented it in the 'JavaClassLookupPayloadBuilder'
https://github.com/apache/beam/blob/c471fd31e67a8c29315e7c1859482177d1fe7afa/sdks/python/apache_beam/transforms/external.py#L270

Reply:
Another idea, rather than having an ignored prefix, is to allow any number of unnamed (empty string) arguments at the beginning which would be considered "positional" arguments.

Reply:
The argument get added as a field name of the schema for the method/constructor. I don't think the field name can be empty for a schema.
Note that we now include a single schema for a method/constructor which simplified things quite a bit and allowed us to support more types in a generic way: https://github.com/apache/beam/blob/919fdef0188f808d1d460bf26c1ca7db1db8648b/model/pipeline/src/main/proto/external_transforms.proto#L102

Reply:
"we can only perform validation if Java reflection preserved arg names" is what concerns me. Is this the default mode? If so, maybe that's good enough. 

Good point about schemas requiring names.

Reply:
It's not available by default. Preserving parameter names requires specifying the Java compiler parameter "-parameters". Beam sets this by default. 

I don't think this is a major issue though. 

We use parameter names to avoid any ambiguities when mapping a Python method/constructor invocation to Java side. If available information is inadequate to unambiguously determine the constructor/method we fail.

https://github.com/apache/beam/blob/587f58e1a1c3f5caa14f9b47144b5a5296cff9b5/sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/JavaClassLookupTransformProvider.java#L238

To perform the actual invocation, we just use the values (similar to a local Java invocation) hence if provided value types do not match, that will result in a failure.

https://github.com/apache/beam/blob/587f58e1a1c3f5caa14f9b47144b5a5296cff9b5/sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/JavaClassLookupTransformProvider.java#L168

Reply:
OK, if the parameter names are not available, let's fail if trying to use named parameters rather than silently hope they match. Types are not enough to disambiguate, e.g. imagine `CopyTransform(String src, String dest)` vs. `CopyTransform(String dest, String src)`. Maybe this'd fail, but something like `ComputeDiffTransform(String newLoc, String prevLoc)` would give the wrong results if someone called `ComputeDiffTransform(prevLoc="old", newLoc="new")`. True one can make this mistake with positional arguments, but at least in that case one is aware that the ordering should be checked.

Reply:
May be we should limit https://github.com/apache/beam/pull/15546 to positional arguments to prevent confusion ? 

I think supporting keyword arguments in "JavaClassLookupPayloadBuilder" for default Java compilation will be useful since any higher level wrappers can easily reorder arguments before calling "JavaClassLookupPayloadBuilder" (to respect its spec).

Also, even when parameter names are available, "JavaClassLookupPayloadBuilder" will still require keyword arguments to be specified in the correct order since Schema (and the Java invocation) requires it.

Reply:
We should accept keyword arguments, for both, only when the parameter names are available. Maybe it makes sense to take this to the list?

Reply:
Will send you a PR on this and we discuss in list as well if needed.

Reply:
cc: @lostluck

Reply:
Sent https://github.com/apache/beam/pull/15557